### PR TITLE
Show Full-Video Labels on thumbnails

### PR DESCRIPTION
**Pull request by mini-bomba**
- [x] I agree to license my contribution under LGPL-3.0 **or** my contribution is from another project with a license compatible with LGPL-3.0

To test this pull request, follow the [instructions in the wiki](https://github.com/ajayyy/SponsorBlock/wiki/Testing-a-Pull-Request).

***
The /videoLabels endpoint was added to SponsorBlockServer quite a while ago, but I didn't see any progress on the extension side, so I decided to try implementing it myself.
Labels are shown as a circle of the category's color with the sponsorblock logo in the middle (screenshot below)
image
When you hover over the icon, it is expanded to a rounded rectangle, with the category name shown.
image
The text color can be white or black, depending on the category color. (this logic was copied from regular fv labels) The logo's color will also update. (yet another screenshot below)
image
Category colors will update live as you change your settings. This is done using CSS variables. (we could probably replace color logic in regular fv labels and the timeline to use those)

Label fetching & API data processing is done on the background page instead of the content script. This allows me to keep a cache of requested hash-prefix chunks to limit the amount of requests. Responses are cached for 1 hour. (or until browser shutdown) 
Requests use hash prefixes. 4 characters seems to be a bit low for labels - at most 2 videos seem to be returned. If hash prefixes are shortened, the amount of requests will be reduced, as this patch caches whole hash-prefix blocks instead of individual videos. I'd even consider having all full-video segments returned in a single request an option - There are only 47681 labels in the DB, and this only needs videoIDs and categories to function.

### CODE REVIEW

**Comment by ajayyy** [comment]
I wonder if it'll look nice slightly translucent

File: src/background.ts [hunk 1]
```diff
@@ -253,3 +257,49 @@ async function sendRequestToCustomServer(type: string, url: string, data = {}) {
 
     return response;
 }
+
+async function getVideoLabel(videoID: VideoID): Promise<Category> {
+    const hashPrefix = (await utils.getHash(videoID, 1)).slice(0, 3);
```
Comment (ajayyy) [inline src/background.ts:262]: If 3 is kept, server needs updates before merging (for cache)

File: src/content.ts [hunk 2]
```diff
@@ -2484,6 +2460,82 @@ function addPageListeners(): void {
...
+}
+
+function setUpThumbnailObserver() {
+    if (thumbnailObserver) {
+        thumbnailObserver.disconnect();
+        thumbnailObserver = null;
+    }
+    if (Config.config.fullVideoSegments) {
+        thumbnailObserver = new MutationObserver(thumbnailObserverHandler);
+        thumbnailObserver.observe(document, {childList: true, subtree: true})
```
Comment (ajayyy) [inline src/content.ts:2473]: This is going to be called for _every_ child list change. I wonder if there's a better option here

Comment (mini-bomba) [reply src/content.ts:2473]: Probably the best option would be to somehow hook into the thumbnail custom element...

**Comment by ajayyy** [comment]
General changes:

- Move cache out of background script. Background script is non persistent on chrome, and will be on all browsers on Manifest v3. This makes it live less long than the browser tab

- Switch document script injection system to work on channel pages by injecting raw html instead of using a URL

- Use mutation listener on thumbails to detect changing related videos and switching from recently uploaded to popular on channels

- Changed cache invalidation

***

I still have a bit to finish before merging, but it looks awesome thanks so much!

**Pull request merged**